elastic/kibana
# Add a welcome screen for new Kibana instances

This addresses [#18828](https://github.com/elastic/kibana/issues/18828).

The welcome screen should show when you view `#/home`, if there are no index patterns. It should allow you to bypass it by heading over to other screens (e.g. `#/management`).  You should also be able to toggle the welcome screen by heading over to `#/management/kibana/settings`, and toggling `Show the welcome screen`. 

This change introduces a delay to the home screen, as we need to query for the existence of data prior to rendering. One thing we might consider doing is adding a max-timeout for that query, and if it hasn't returned in a reasonable time, just show the normal home page. For poor connections, this should improve the experience.

<img width="1353" alt="image" src="https://user-images.githubusercontent.com/833377/43324083-42dcb6aa-9181-11e8-91b1-3862f19b91b8.png">

Clicking "I don't need help" should hide the welcome screen forever. This is done on a per-browser basis via `localStorage`.

Clicking "Play around with sample data" should bring you to the sample data screen.

## Fade in

The welcome screen fades in right now, as there doesn't appear to be a good way for us to keep the loading indicator up until the welcome screen has rendered. (A good way to do that is coming in the new platform.) Without the fade in, there is a bad flicker effect.

## Flaky tests

This is a potential source of flaky tests. :/ The welcome screen will show up after a delay (after it's queried Elasticsearch to determine whether or not this is a new Kibana instance). This means that any tests that expect the old home screen might pass if they run fast enough to happen before the welcome screen shows up. I'm not sure how to fix this. We *could* disable the welcome screen by default for test scenarios, and then enable it only for explicit welcome screen tests.

Alternatively, we could hack the UI on the home screen so that the Kibana loading indicator shows up until we've done our query, then we'd show the welcome screen or the home screen, depending on the result of the query. This would have a secondary benefit of avoiding the flash / blink effect caused by the page rendering, then a bit later the welcome screen showing up. I don't know of a good way to accomplish this without resorting to serious hacks, though, as I don't see a good way to tell Kibana to keep the loading screen up until the current app says to remove it.

## Global settings

Another note for review is that I'm using a combination of localStorage and the uiSettingsClient to save / load whether or not the welcome screen has been dismissed. This means that if one user dismisses the welcome screen, it's dismissed only for their browser. Incognito or another browser will show the welcome screen again. I'm using a setting in uiSettingsClient to enable the welcome screen to be disabled system-wide. Is uiSettingsClient the right tool for that job?

## Testing

- Spin up a new ELK instance.
- Load Kibana
- You should see the welcome screen
- Try disabling the welcome screen in `#/management/kibana/settings`
- Try enabling it again
- Create sample data by clicking "Play round with sample data" and then clicking "Add" in the sample data card.
- Confirm that once the data has been generated, you no longer see the welcome screen when hitting the home page
- Clear elasticsearch (e.g. restart it in dev mode)
- Reload `#/home`
- You should see the welcome screen
- Create some data `node scripts/makelogs`
- Head over to `#/management` and define an index pattern
- You should no longer see the welcome screen when you view `#/home`
